src/core/ipfs/object.js

```diff
@@ -1,141 +1,247 @@
 'use strict'
 
-const Block = require('ipfs-block')
 const mDAG = require('ipfs-merkle-dag')
+const waterfall = require('run-waterfall')
+const promisify = require('promisify-es6')
+const bs58 = require('bs58')
 const DAGNode = mDAG.DAGNode
+const DAGLink = mDAG.DAGLink
+
+function normalizeMulithash (multihash, enc) {
+  if (typeof multihash === 'string') {
+    if (enc === 'base58') {
+      return multihash
+    }
+
+    return new Buffer(multihash, enc)
+  } else if (Buffer.isBuffer(multihash)) {
+    return multihash
+  } else {
+    throw new Error('unsupported multihash')
+  }
+}
+
+function parseBuffer (buf, encoding) {
+  switch (encoding) {
+    case 'json':
+      return parseJSONBuffer(buf)
+    case 'protobuf':
+      return parseProtoBuffer(buf)
+    default:
+      throw new Error(`unkown encoding: ${encoding}`)
+  }
+}
+
+function parseJSONBuffer (buf) {
+  try {
+    const parsed = JSON.parse(buf.toString())
+    const links = (parsed.Links || []).map((link) => {
+      return new DAGLink(
+        link.Name,
+        link.Size,
+        new Buffer(bs58.decode(link.Hash))
+      )
+    })
+    return new DAGNode(new Buffer(parsed.Data), links)
+  } catch (err) {
+    throw new Error('failed to parse JSON: ' + err)
+  }
+}
+
+function parseProtoBuffer (buf) {
+  const node = new DAGNode()
+  node.unMarshal(buf)
+  return node
+}
 
 module.exports = function object (self) {
-  return {
-    new: (template, callback) => {
-      if (!callback) {
-        callback = template
+  function editAndSafe (edit) {
+    return (multihash, options, cb) => {
```

Comment:
A good pattern is that each time there is a top level callback, calling it `callback`, so that the inner callbacks (scoped) are named just `cb`. Makes it easier to see when we are calling to out of the function or inside function operations (also, sometimes saves time in debugging)